### ISSUE TITLE
Front Javascript support

### ISSUE THREAD

**Issue by vykimo** [porsager/postgres#53]
Hi,
I discovered that `postgres` is not supported by front-end javascript. I mean the real vanilla javascript (not the node.js one).
I am asking it because I'd have liked to use it in a Cloudflare Service Worker.
That's too bad because `postgres` seems really promising. In the aspect of lightweight (small webpack bundle size), the fastest (interesting for small execution time) and few dependencies (it looks like only few lib are making trouble, fs, etc..)

What do you think?

**Comment by porsager** [comment]
It's a very interesting thought, but `fs` is not the real issue. You will need to have a transport to the database other than tcp, which eventually means a server in between your Cloudflare workers and the Postgres database. 

If you want to dig deeper I think the most straight forward way would be to have a tcp over http tunnel and then mock the `net` module to use that tunnel in your Cloudflare worker.

It should be fairly straight forward to transpile this library using eg. rollup and replace the node module references to eg. net with this mock for the tcp over http tunnel, produce an esm module, and then use that from within your workers.

With that said I don't think it's the most optimal solution speaking the postgres protocol over http. It'll be a lot more requests than having a server in the middle that can keep persistent connections to the database and handle queries.

**Comment by vykimo** [comment]
Thank you for your answer, so for you a solution like this is the only possible way to deal with postgres db : https://github.com/cloudflare/db-connect
A HTTP tunnel... So sad that TCP is not supported in cloudflare.

**Comment by Minigugus** [comment]
I agree with @vykimo, a negative point for `postgres` is its reliance on the NodeJS builtins. I don't mean that `postgres` should run directly in the browser, but that others standalone JavaScript runtimes like [Deno](https://deno.land/) or [QuickJS](https://bellard.org/quickjs/) should be able to run this lib as they support native TCP connections and other back-end features.

I know it's difficult to rewrite an entire library this way but maybe one solution could be to use dependency injection, with only default support for the Node :

```js
// The default configuration
postgres({
  env: {

    // Simple but NodeJS-oriented
    fs: require('fs'),
    net: require('net'),
    // ...

    // Complex but plateform-agnostic (permit optionnal features, easier to adapt to another environment)
    createServer: require('net').createServer,
    MD5Digest: () => { /* ... */ },
    // ...
  },
  // ...
});
```

**Comment by porsager** [comment]
> A HTTP tunnel... So sad that TCP is not supported in cloudflare.

Even if tcp was supported it would be silly to speak the postgres protocol directly from a worker because they are short lived and single request only. That would mean you would have to authenticate with the db for every single request and have no way to do connection pooling and queueing which is bad for postgres as it doesn't work well with a lot of concurrent connections. 

The best way to do this is still having a server in the middle to do those things. It would be really cool though to mimic the Postgres.js api in a client library that would then speak with this server over http.

That's certainly true @Minigugus, would be very neat to  and as I see it not even that hard to achieve. The biggest part is an alternative to the node only Buffer type.

> I agree with @vykimo, a negative point for postgres is its reliance on the NodeJS builtins.

The node builtins are fairly simple so making mocks to support other things than node should be fairly straight forward. I think it might be better to make transpiled versions with these mocks for other environments rather than dependency injection. 

I have been looking at Deno, but it's just not near node performance wise yet, and it still doesn't have tls support (last time i checked at least).

> I know it's difficult to rewrite an entire library this way but maybe one solution could be to use dependency injection, with only default support for the Node.

I actually think it'd be at most a one day job, but currently I don't see any serious use cases outside Node.js.

**Comment by fjf2002** [comment]
(Not exactly this issue's point, but I am crazy enough to run this postgres library inside the browser, using net-browserify - if anyone is interested)

**Comment by vykimo** [comment]
> (Not exactly this issue's point, but I am crazy enough to run this postgres library inside the browser, using net-browserify as polyfill for the nodejs net module - if anyone is interested. Despite the naming, the latter library also works with webpack.)

I am interested and curious how a tcp connection is possible without any tunnel.

**Comment by fjf2002** [comment]
> I am interested and curious how a tcp connection is possible without any tunnel.

Well, net-browserify _is_ a tunnel. It does tcp-over-websocket and consists of two components: the client-side nodejs-net-interface and a server-side process that forwards the wrapped tcp connection to an arbitrary server.

**Comment by porsager** [comment]
Yeah, it definitely has it's use cases, but it seems making it work is already possible by transpiling with specific mocks 🙂

**Comment by qimingweng** [comment]
@porsager I think deno has TLS support now, https://github.com/denoland/deno/issues/1568, and also is approaching 1.0 this week. (https://blog.logrocket.com/deno-1-0-what-you-need-to-know/)

Your goal of 0 dependencies and clean surface area seem quite aligned with the deno project, even if it isn't going to be the most popular use case. In fact, since node-postgres (pg) won't have an easy time switching over, you could become the defacto postgres library.

Personally, deno is kind of awesome for writing short scripts, and sometimes having the ability to query a server is useful.

PS. I was hoping, since the package name was originally from https://github.com/ry, that this would be the most sanctioned postgres library!

**Comment by jgod** [comment]
Without opening a new issue, and to add visibility to the +1, I am writing this comment to say that Deno support would be very spicy.

**Comment by Minigugus** [comment]
@jgod Spicy, in fact :smile: 
https://github.com/Minigugus/postgres-deno :wink:

**Comment by porsager** [comment]
So awesome @Minigugus ! :D 

I wanted to do the same, but not until I had solved the last api issues and new features around safe dynamic sql queries, so really neat that you've already got this far!! 

My plan was to make this package support Deno directly with the few mocks needed, so if you want to collaborate on that here, that'd be cool.

**Comment by Minigugus** [comment]
@porsager Thanks :smile: 

> My plan was to make this package support Deno directly with the few mocks

Most mocks are related to `Buffer`s, so I just swtiched to `Uint8Array`s :wink:

However, Deno only support ES modules but your code is currently CommonJS, si if you want to support Deno directly, a build step will be required for compatibility :confused: 

Also, there is no good `setImmediate` alternative in Deno yet (no `postMessage` to polyfill it), which means that some optimizations/features cannot be ported properly ([buffering](https://github.com/porsager/postgres/blob/5e3fff861a8e6ace844dba50b32d78e057c18e57/lib/connection.js#L298-L300) for instance)

Anyway, it's working! And the Deno support is quite [promising regarding performance](https://doc.deno.land/https/github.com/denoland/deno/releases/latest/download/lib.deno.d.ts#Deno.iter) :smile:

**Comment by porsager** [comment]
Hey @Minigugus .. Are you on gitter? I've just created a chat room for Postgres.js there.. You can also do DM chats there, maybe it'd be easier to discuss deno support there? https://gitter.im/porsager/postgres
